### PR TITLE
ascii data plugin for boundary traction

### PR DESCRIPTION
**Pull request by jdannberg**
This pull request adds the option to specify pressure data along the boundaries of a model in form of an ascii data file that will be used as traction boundary conditions (the pressure given in the data file is applied as traction normal to the surface of a given boundary).

This is useful for models where it is difficult to prescribe the traction as a function such as necessary for the `function` plugin. Instead, models can be run once with approximate values for the boundary traction, and the solution of such a test model can be used to extract the pressure values on a profile close to the boundary and use them as traction boundary conditions for the "real" model (if necessary, several iterations can be done). 
This is a bit cumbersome, as the pressure profile has to be extracted by hand each time, but it was the only good way I found that made it possible to specify a boundary traction without much in- or outflow. In addition, the approach should also work for time-dependent traction.

### CODE REVIEW

**Comment by jdannberg** [comment]
@anne-glerum: Is this something that could be useful for you/you already thought about? I remember you wanted to use traction boundary conditions at some point, and as I started to use them now, I found it extremely difficult to specify the traction in a way that I had no artificial in- or outflow.

**Comment by bangerth** [comment]
@jdannberg -- if you want no inflow or outflow, why not just prescribe tangential boundary conditions? That should have exactly the effect you are looking for!

**Comment by jdannberg** [comment]
I do want in- and outflow, but according to the dynamic pressure in the model. With the current traction boundary conditions it is extremely difficult to prescribe the traction in a way so that the dynamic pressure at the model boundary is the same as the dynamic pressure in an undisturbed mantle close to the boundary; in most cases I got in- and outflow velocities of several 10s or 100s of meters/yr. 

In my case, for example, I want close to no in- and outflow at the beginning of the model, but as a plume develops over time, plume material should be able to flow out at one of the side boundaries. I am trying to reproduce a model from the literature, and I think this kind of boundary is often used in geodynamics (and it's often called 'open bounday'). 

First, I though it would be enough to prescribe the adiabatic pressure at the boundary, but that does not take into account thermal boundary layers, or any lateral heterogeneities (my upper thermal boundary layer also varies in thickness), so I got quite large in- and outflow velocities already in the very first time step.

**Comment by anne-glerum** [comment]
@jdannberg: That's interesting indeed. I am using open boundaries at the moment and wrote two plugins for it. The first calculates a pressure profile in a specific point based on the initial conditions for temperature and composition, much like the initial adiabatic profile is calculated. This is sufficient if you have no lateral variations along a boundary or changes in lithostatic pressure over time. The second plugin divides the boundaries in bins and extracts values from the previous solution needed to compute the lithostatic pressure in each bin. From this grid the requested pressure in a specific point is interpolated. This plugin is updated every timestep, and this seems to work for the 2D benchmarks and subduction models we've tested it on. I wanted to discuss this during the hackaton, but you beat me to it :)

**Comment by bangerth** [comment]
@jdannberg -- I see. Yes, I've always wondered how people do that. Just having a floating continent that's not rising up or sinking down has always struck me as a challenge. What @anne-glerum does seems like a reasonable strategy.

**Comment by naliboff** [comment]
@jdannberg - open boundaries are quite tricky and I've never fully understood myself what many studies are actually doing when they say 'open boundary'.  As per the discussion above, there has to be some constraint on the pressure (or more generally stress field) in order to get a reasonable solution.   This comes back to what people do when a 'winkler boundary condition' is prescribed at the base of a model.  For reference, could you point to the specific paper your are looking at reproducing?

**Comment by jdannberg** [comment]
@anne-glerum: It's great to hear you found a solution for your models! And we can still discuss it at the hackathon. To me it sounds like all of these plugins could be useful for different application cases, so there's no reason why we shouldn't have all these different options in Aspect. 

@naliboff: For me it is not important to get exactly the same solution as in the paper (also because it has a visco-elasto-plastic rheology anyway), but to get a similar model setup and start from there. It is the Sobolev at al. (2011) paper (http://www.nature.com/nature/journal/v477/n7364/abs/nature10385.html), so there is a plume rising from the bottom of the model boundary, and the left model boundary is open. So for me it will probably be enough to use one pressure profile during the whole model evolution, but at some point I plan to extend the model to three dimensions, and it will probably vary in lithospheric thickness along the boundary, so I wanted a model that is flexible enough to do that (but I agree that @anne-glerum's approach with extracting the solution from the previous time step and using that to compute the new lithostatic pressure would work very well for that too). 

But independent of what I wanted to use the model for, I thought it would be a good addition to Aspect. Other people may find different applications for it, and we have ascii data models for all of the other boundary conditions.

**Comment by bangerth** [comment]
Yes, I agree that a plugin that computes the stress from some "average" depth dependent pressure would make great sense.

**Comment by naliboff** [comment]
@jdannberg - I think both your and @anne-glerum plugins approach will be quite useful in specific cases.    To minimize large unrealistic velocities at the boundary in your model, especially in a 3D scenario, I think   the ideal procedure would be to apply the pressure from the prior solution along a given prior as a traction in the current solution.  As long as the non-lithostatic pressure is not changing dramatically between time-steps, though, an average pressure profile would hopefully work.  Tricky stuff, I've been discussing a similar issue with someone else in the context of regional stress modeling (different code)!  Looking forward to seeing how the different methods turn out!

**Comment by anne-glerum** [comment]
@jdannberg - I'm now testing the second plugin for 3D cases with varying lithospheric thickness along the boundaries, so I can show the results at the hackaton. 

Glad to hear the plugins seem useful, I'll do a pull request before/during the hackaton.
We also devised some 2D and 3D tests, perhaps we can compare results for the different methods as well then. Would that be interesting for the manual?

File: source/traction_boundary_conditions/ascii_data.cc [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                                                 "files need to be sorted in a specific order: "
+                                                 "the first coordinate needs to ascend first, "
+                                                 "followed by the second in order to "
+                                                 "assign the correct data to the prescribed coordinates. "
+                                                 "If you use a spherical model, "
+                                                 "then the data will still be handled as Cartesian, "
+                                                 "however the assumed grid changes. 'x' will be replaced by "
+                                                 "the radial distance of the point to the bottom of the model, "
+                                                 "'y' by the azimuth angle and 'z' by the polar angle measured "
+                                                 "positive from the north pole. The grid will be assumed to be "
```
Comment (tjhei) [inline source/traction_boundary_conditions/ascii_data.cc]: I assume this comment is copied, right? I think the radius "r" is never a column in the file. Right?

File: tests/traction_ascii_data.prm [hunk 2]
```diff
@@ -0,0 +1,134 @@
+# Listing of Parameters
```
Comment (tjhei) [inline tests/traction_ascii_data.prm]: can you add a short description please?

**Comment by tjhei** [comment]
Looks good to me, but can you please address the minor comments?

File: source/traction_boundary_conditions/ascii_data.cc (hunk 1 above)
Comment (jdannberg) [inline source/traction_boundary_conditions/ascii_data.cc]: Why do you think the radius can't be a column? If it is a spherical shell geometry model with an opening angle of <360°, we also need boundary conditions at the sides, see for example in `data/velocity-boundary-conditions/ascii-data/test/shell_3d_east.0.txt`

Comment (tjhei) [inline source/traction_boundary_conditions/ascii_data.cc]: Ah, so I guess there is a lot more magic going on in the AsciiData thing than I thought.

**Comment by tjhei** [comment]
Changelog entry? And please merge yourself if you see green lights. :-)

**Comment by jdannberg** [comment]
@anne-glerum - Yes, I think it's a good idea to discuss that at the hackathon, and also to look at the differences between the different approaches. And it would be great to have a cookbook with an application for the traction boundary conditions for the manual, so if we find it makes a difference which plugin is used, we can certainly include that.
